util/src/main/java/com/google/cloud/hadoop/util/RetryHttpInitializer.java

```diff
@@ -151,30 +158,38 @@ private static class UnsuccessfulResponseHandler implements HttpUnsuccessfulResp
 
     private final HttpCredentialsAdapter credentials;
     private final HttpBackOffUnsuccessfulResponseHandler delegate;
+    private final RequestTracker tracker;
 
-    public UnsuccessfulResponseHandler(HttpCredentialsAdapter credentials) {
+    public UnsuccessfulResponseHandler(HttpCredentialsAdapter credentials, RequestTracker tracker) {
       this.credentials = credentials;
       this.delegate =
           new HttpBackOffUnsuccessfulResponseHandler(BACKOFF_BUILDER.build())
               .setBackOffRequired(BACK_OFF_REQUIRED);
+      this.tracker = tracker;
     }
 
     @Override
     public boolean handleResponse(HttpRequest request, HttpResponse response, boolean supportsRetry)
         throws IOException {
       logResponseCode(request, response);
+      tracker.trackUnsuccessfulResponseHandler(response);
 
       if (credentials != null && credentials.handleResponse(request, response, supportsRetry)) {
         // If credentials decides it can handle it, the return code or message indicated something
         // specific to authentication, and no backoff is desired.
         return true;
       }
 
+      long backOffStartTime = System.currentTimeMillis();
       if (delegate.handleResponse(request, response, supportsRetry)) {
+        tracker.trackBackOffCompleted(backOffStartTime, request);
         // Otherwise, we defer to the judgement of our internal backoff handler.
+        tracker.trackRetryStarted();
```

Comment:
I would recommend overriding `HttpBackOffUnsuccessfulResponseHandler` handleResponse and adding this logic there instead of polluting `UnsuccessfulResponseHandler`'s `handleResponse` which has nothing to do with the retries.

Reply:
ok. I get what you are saying. Agree, that would have worked as well. The current implementation is fine too.